nextest-runner/src/test_filter.rs

```diff
@@ -63,30 +63,240 @@ enum TestFilterExprs {
+    }
+
+    /// Adds an exact pattern to the set of patterns.
+    pub fn add_exact_pattern(&mut self, pattern: String) {
+        match self {
+            Self::SkipOnly { skip_patterns } => {
+                *self = Self::Patterns {
+                    patterns: Vec::new(),
+                    exact_patterns: [pattern].into_iter().collect(),
+                    skip_patterns: mem::take(skip_patterns),
+                };
+            }
+            Self::Patterns { exact_patterns, .. } => {
+                exact_patterns.insert(pattern);
+            }
+        }
+    }
+
+    /// Adds a skip pattern to the set of patterns.
+    pub fn add_skip_pattern(&mut self, pattern: String) {
+        match self {
+            Self::SkipOnly { skip_patterns } => {
+                skip_patterns.push(pattern);
+            }
+            Self::Patterns { skip_patterns, .. } => {
+                skip_patterns.push(pattern);
+            }
+        }
+    }
+
+    fn resolve(self) -> Result<ResolvedFilterPatterns, TestFilterBuilderError> {
+        match self {
+            Self::SkipOnly { mut skip_patterns } => {
+                if skip_patterns.is_empty() {
+                    Ok(ResolvedFilterPatterns::All)
+                } else {
+                    // sort_unstable allows the PartialEq implementation to work correctly.
+                    skip_patterns.sort_unstable();
+                    let skip_pattern_matcher = Box::new(AhoCorasick::new(&skip_patterns)?);
+                    Ok(ResolvedFilterPatterns::SkipOnly {
+                        skip_patterns,
+                        skip_pattern_matcher,
+                    })
+                }
+            }
+            Self::Patterns {
+                mut patterns,
+                exact_patterns,
+                mut skip_patterns,
+            } => {
+                // sort_unstable allows the PartialEq implementation to work correctly.
+                patterns.sort_unstable();
+                skip_patterns.sort_unstable();
+
+                let pattern_matcher = Box::new(AhoCorasick::new(&patterns)?);
+                let skip_pattern_matcher = Box::new(AhoCorasick::new(&skip_patterns)?);
+
+                Ok(ResolvedFilterPatterns::Patterns {
+                    patterns,
+                    exact_patterns,
+                    skip_patterns,
+                    pattern_matcher,
+                    skip_pattern_matcher,
+                })
+            }
+        }
+    }
+}
+
 #[derive(Clone, Debug)]
-enum NameMatch {
-    EmptyPatterns,
-    MatchSet {
+enum ResolvedFilterPatterns {
+    All,
+    SkipOnly {
+        skip_patterns: Vec<String>,
+        skip_pattern_matcher: Box<AhoCorasick>,
+    },
+    Patterns {
         patterns: Vec<String>,
-        matcher: Box<AhoCorasick>,
+        exact_patterns: HashSet<String>,
+        skip_patterns: Vec<String>,
+        pattern_matcher: Box<AhoCorasick>,
+        skip_pattern_matcher: Box<AhoCorasick>,
     },
```

Comment:
This is the resolved form with matchers for all the substring patterns. I separated `All` here mostly for convenience.